eventdisplay/FairXMLNode.cxx

```diff
@@ -8,232 +8,105 @@
-			TString temp =(( FairXMLAttrib*)fAttrib->At(i))->GetName();
+			fAttrib.Add(new FairXMLAttrib(atrib->GetName(),atrib->GetValue()));
 		}
-		fAttrib->Add(new FairXMLAttrib(name,value));
-		fNAttrib++;
 	}
 }
 
-Int_t FairXMLNode::GetNChildren() const {
-	if(fMode == kModeNull){
-		//FairCout::Error("dont call for null node");
-		return 0;
+FairXMLNode::FairXMLNode(const FairXMLNode& other):
+		FairXMLNode(other.GetName(),other.GetValue()){
+	for(int i=0;i<other.fChildren.GetEntries();i++){
+		fChildren.Add(new FairXMLNode(*other.GetChild(i)));
 	}
-	return fNChildren;
-}
-
-Int_t FairXMLNode::GetNAttributes() const {
-	if(fMode == kModeNull){
-		//FairCout::Error("not for null node");
-		return 0;
-	}
-	return fNAttrib;
-}
-
-TList* FairXMLNode::GetChildren() const {
-	if(fMode == kModeNull){
-		//FairCout::Error("no child in null node");
-		return NULL;
+	for(int i=0;i<other.fAttrib.GetEntries();i++){
+		fAttrib.Add(new FairXMLAttrib(*other.GetAttrib(i)));
 	}
-	return fChildren;
 }
 
-FairXMLNode * FairXMLNode::fgNullInstance = NULL;
-
 FairXMLNode* FairXMLNode::GetChild(TString name, Int_t count) const{
-	if(fMode == kModeNull){
-		//FairCout::Error("no child in null node");
-		return FairXMLNode::NullInstance();
-	}
-	if(fNChildren==0){
-		//FairCout::Error("No children in this node");
-		return FairXMLNode::NullInstance();
-	}
 	Int_t control_index = 0;
-	for(int i=0;i<fChildren->GetEntries();i++){
-		FairXMLNode *node = (FairXMLNode*)fChildren->At(i);
+	for(int i=0;i<fChildren.GetEntries();i++){
+		FairXMLNode *node = GetChild(i);
 		TString temp = node->GetName();
 		if(temp==name){
 			control_index++;
 		}
 		if(control_index>count)
 			return node;
 	}
-	//FairCout::Error("No child found by name");
-	return FairXMLNode::NullInstance();
-}
-
-FairXMLNode* FairXMLNode::GetChild(Int_t index) const{
-	if(fMode == kModeNull){
-		return FairXMLNode::NullInstance();
-	}
-	if(fNChildren>index){
-		return (FairXMLNode*)fChildren->At(index);
-	}else{
-		return FairXMLNode::NullInstance();
-	}
-}
-
-Bool_t FairXMLNode::IsImport() const{
-	if(fMode==kModeImport) return kTRUE;
-	return kFALSE;
-}
-
-Bool_t FairXMLNode::IsNull() const {
-	if(fMode == kModeNull) return kTRUE;
-	return kFALSE;
-}
-
-FairXMLNode* FairXMLNode::NullInstance() {
-	if(fgNullInstance){
-		return fgNullInstance;
-	}
-	fgNullInstance = new FairXMLNode();
-	return fgNullInstance;
-}
-
-FairXMLNode::FairXMLNode() {
-	fName = "null";
-	fValue= "null";
-	fChildren = NULL;
-	fAttrib = NULL;
-	fNChildren = 0;
-	fNAttrib = 0;
-	fMode = kModeNull;
+	return nullptr;
 }
 
-void FairXMLNode::SetValue(TString value) {
-	if(fMode==kModeExport){
-		fValue = value;
+FairXMLAttrib* FairXMLNode::GetAttrib(TString name) const{
+	Int_t control_index = 0;
+	for(int i=0;i<fAttrib.GetEntries();i++){
+		FairXMLAttrib *node = static_cast<FairXMLAttrib*>(fAttrib.At(i));
+		TString temp = node->GetName();
+		if(temp==name){
+			return node;
+		}
 	}
+	return nullptr;
 }
```

Comment:
This looks like a reimplementation of [`TObject * TList::FindObject (const char * name) const`](https://root.cern.ch/doc/master/classTList.html#a85c8734ab131c0db5f68c724e47496a7). So, `return static_cast<FairXMLAttrib*>(fAttrib.FindObject(name.Data()));` or similar should do the job.

Reply:
Fixed, this "Delete a TObjLink object. " in ROOT documentation discourages from using this function.
I also removed GetNAttributes(TString name). I found out that according to XML spec we should not have two attributes with the same name in same element.

Reply:
> this "Delete a TObjLink object. " in ROOT documentation discourages from using this function.

Didn't even see that. Ya, it's a bug in the generation of their docs, the line comes from the comment of [another function](https://github.com/root-project/root/blob/331efa4c00fefc38980eaaf7b41b8e95fcd1a23b/core/cont/src/TList.cxx#L555-L592).

Reply:
Submitted a [fix](https://github.com/root-project/root/pull/3421).